### ISSUE TITLE
needless_match in nightly (2022-03-14) ignores enum arguments

### ISSUE THREAD

**Issue by vakaras** [rust-lang/rust-clippy#8542]
### Summary

When a matched enum has multiple arguments and one of them is replaced, the `needless_match` lint still reports an error even though the match is not needless.

### Lint Name

needless_match

### Reproducer

I tried this code:

```rust
pub enum E {
    Variant1(u32, bool),
    Variant2(u32, bool),
}

impl E {
    pub fn set_b(self, b: bool) -> Self {
        match self {
            E::Variant1(a, _) => E::Variant1(a, b),
            E::Variant2(a, _) => E::Variant2(a, b),
        }
    }
}
```

I saw this happen:

```
> cargo clippy
    Checking needless_match v0.1.0 (/tmp/bla/needless_match)
warning: this match expression is unnecessary
  --> src/lib.rs:9:9
   |
9  | /         match self {
10 | |             E::Variant1(a, _) => E::Variant1(a, b),
11 | |             E::Variant2(a, _) => E::Variant2(a, b),
12 | |         }
   | |_________^ help: replace it with: `self`
   |
   = note: `#[warn(clippy::needless_match)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_match

warning: `needless_match` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s

```

I expected to see this happen: no warning should be emitted because the new value is constructed with a new `b`.

### Version

```text
rustc 1.61.0-nightly (285fa7ecd 2022-03-14)
binary: rustc
commit-hash: 285fa7ecd05dcbfdaf2faaf20400f5f92b39b3c6
commit-date: 2022-03-14
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by teor2345** [comment]
The lint also triggers on nightly-2022-03-14 for code like this ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=cbbd0259c8b3396ea27c56c119fc64c9)):

```rust
 enum Resultt<T, E> {
     Okk(T),
     Errr(E),
 }
 
 use Resultt::*;
 
 impl<T, E> Resultt<T, E> {
    pub const fn as_ref(self: &Resultt<T, E>) -> Resultt<&T, &E> {
        match self {
            Okk(x) => Okk(x),
            Errr(x) => Errr(x),
        }
    }
}
```

`std::Result` [does it slightly differently](https://doc.rust-lang.org/src/core/result.rs.html#643-648), and doesn't trigger the lint:
```rust
    pub const fn as_ref(self: &Result<T, E>) -> Result<&T, &E> {
        match *self {
            Ok(ref x) => Ok(x),
            Err(ref x) => Err(x),
        }
    }
```

(I've desugared `&self`, to make it clear what the type change is.)
